// interweave/lib/constants.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ATTRIBUTES_TO_PROPS = exports.ATTRIBUTES = exports.FILTER_CAST_BOOL = exports.FILTER_CAST_NUMBER = exports.FILTER_DENY = exports.FILTER_ALLOW = exports.TAGS_BLACKLIST = exports.TAGS = exports.CONFIG_BLOCK = exports.CONFIG_INLINE = exports.TYPE_INLINE_BLOCK = exports.TYPE_BLOCK = exports.TYPE_INLINE = exports.PARSER_DENY = exports.PARSER_ALLOW = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var PARSER_ALLOW = 1;
exports.PARSER_ALLOW = PARSER_ALLOW;
var PARSER_DENY = 2;
exports.PARSER_DENY = PARSER_DENY;
var TYPE_INLINE = 'inline';
exports.TYPE_INLINE = TYPE_INLINE;
var TYPE_BLOCK = 'block';
exports.TYPE_BLOCK = TYPE_BLOCK;
var TYPE_INLINE_BLOCK = 'inline-block';
exports.TYPE_INLINE_BLOCK = TYPE_INLINE_BLOCK;
var CONFIG_INLINE = {
  rule: PARSER_ALLOW,
  type: TYPE_INLINE,
  inline: true,
  block: false,
  self: false,
  void: false,
  parent: [],
  children: []
};
exports.CONFIG_INLINE = CONFIG_INLINE;
var CONFIG_BLOCK = {
  rule: PARSER_ALLOW,
  type: TYPE_BLOCK,
  inline: true,
  block: true,
  self: true,
  void: false,
  parent: [],
  children: []
};
exports.CONFIG_BLOCK = CONFIG_BLOCK;
var tagConfigs = {
  a: {
    type: TYPE_INLINE_BLOCK,
    block: true
  },
  address: {
    self: false
  },
  audio: {
    children: ['track', 'source']
  },
  br: {
    inline: false,
    void: true
  },
  button: {
    type: TYPE_INLINE_BLOCK
  },
  dd: {
    parent: ['dl']
  },
  dl: {
    children: ['dt']
  },
  dt: {
    parent: ['dl'],
    children: ['dd']
  },
  figcaption: {
    parent: ['figure']
  },
  footer: {
    self: false
  },
  header: {
    self: false
  },
  h1: {
    self: false
  },
  h2: {
    self: false
  },
  h3: {
    self: false
  },
  h4: {
    self: false
  },
  h5: {
    self: false
  },
  h6: {
    self: false
  },
  hr: {
    inline: false,
    block: false,
    void: true
  },
  img: {
    inline: false,
    void: true
  },
  li: {
    self: false,
    parent: ['ul', 'ol']
  },
  main: {
    self: false
  },
  ol: {
    children: ['li']
  },
  picture: {
    children: ['source', 'img']
  },
  source: {
    inline: false,
    parent: ['audio', 'video', 'picture'],
    void: true
  },
  span: {
    self: true
  },
  table: {
    children: ['thead', 'tbody', 'tfoot', 'tr']
  },
  tbody: {
    parent: ['table'],
    children: ['tr']
  },
  td: {
    parent: ['tr']
  },
  tfoot: {
    parent: ['table'],
    children: ['tr']
  },
  th: {
    parent: ['tr']
  },
  thead: {
    parent: ['table'],
    children: ['tr']
  },
  tr: {
    parent: ['table', 'tbody', 'thead', 'tfoot'],
    children: ['th', 'td']
  },
  track: {
    inline: false,
    parent: ['audio', 'video'],
    void: true
  },
  ul: {
    children: ['li']
  },
  video: {
    children: ['track', 'source']
  }
};

function createConfigBuilder(config) {
  return function (tagName) {
    tagConfigs[tagName] = (0, _objectSpread2.default)({}, config, tagConfigs[tagName]);
  };
}

['a', 'abbr', 'b', 'br', 'button', 'canvas', 'cite', 'code', 'del', 'dfn', 'em', 'i', 'iframe', 'img', 'ins', 'kbd', 'label', 'mark', 'output', 'picture', 'q', 's', 'samp', 'script', 'source', 'span', 'strong', 'sub', 'sup', 'style', 'time', 'track', 'u', 'var', 'video'].forEach(createConfigBuilder(CONFIG_INLINE));
['address', 'article', 'aside', 'audio', 'blockquote', 'dd', 'details', 'div', 'dl', 'dt', 'fieldset', 'figcaption', 'figure', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hr', 'legend', 'li', 'main', 'nav', 'ol', 'p', 'pre', 'section', 'summary', 'table', 'thead', 'tbody', 'tfoot', 'tr', 'th', 'td', 'ul'].forEach(createConfigBuilder(CONFIG_BLOCK));
var TAGS = Object.freeze(tagConfigs);
exports.TAGS = TAGS;
var TAGS_BLACKLIST = {
  applet: true,
  body: true,
  canvas: true,
  embed: true,
  frame: true,
  frameset: true,
  head: true,
  html: true,
  iframe: true,
  object: true,
  script: true,
  style: true
};
exports.TAGS_BLACKLIST = TAGS_BLACKLIST;
var FILTER_ALLOW = 1;
exports.FILTER_ALLOW = FILTER_ALLOW;
var FILTER_DENY = 2;
exports.FILTER_DENY = FILTER_DENY;
var FILTER_CAST_NUMBER = 3;
exports.FILTER_CAST_NUMBER = FILTER_CAST_NUMBER;
var FILTER_CAST_BOOL = 4;
exports.FILTER_CAST_BOOL = FILTER_CAST_BOOL;
var ATTRIBUTES = Object.freeze({
  alt: FILTER_ALLOW,
  cite: FILTER_ALLOW,
  class: FILTER_ALLOW,
  colspan: FILTER_CAST_NUMBER,
  controls: FILTER_CAST_BOOL,
  datetime: FILTER_ALLOW,
  default: FILTER_CAST_BOOL,
  disabled: FILTER_CAST_BOOL,
  dir: FILTER_ALLOW,
  height: FILTER_ALLOW,
  href: FILTER_ALLOW,
  id: FILTER_ALLOW,
  kind: FILTER_ALLOW,
  label: FILTER_ALLOW,
  lang: FILTER_ALLOW,
  loop: FILTER_CAST_BOOL,
  muted: FILTER_CAST_BOOL,
  poster: FILTER_ALLOW,
  role: FILTER_ALLOW,
  rowspan: FILTER_CAST_NUMBER,
  sizes: FILTER_ALLOW,
  span: FILTER_CAST_NUMBER,
  src: FILTER_ALLOW,
  srcset: FILTER_ALLOW,
  target: FILTER_ALLOW,
  title: FILTER_ALLOW,
  type: FILTER_ALLOW,
  width: FILTER_ALLOW
});
exports.ATTRIBUTES = ATTRIBUTES;
var ATTRIBUTES_TO_PROPS = Object.freeze({
  class: 'className',
  colspan: 'colSpan',
  datetime: 'dateTime',
  rowspan: 'rowSpan',
  srcset: 'srcSet'
});
exports.ATTRIBUTES_TO_PROPS = ATTRIBUTES_TO_PROPS;
