// interweave/lib/Markup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Element = _interopRequireDefault(require("./Element"));

var _Parser = _interopRequireDefault(require("./Parser"));

var Markup = function (_React$PureComponent) {
  (0, _inherits2.default)(Markup, _React$PureComponent);

  function Markup() {
    (0, _classCallCheck2.default)(this, Markup);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Markup).apply(this, arguments));
  }

  (0, _createClass2.default)(Markup, [{
    key: "getContent",
    value: function getContent() {
      var _this$props = this.props,
          _this$props$content = _this$props.content,
          content = _this$props$content === void 0 ? '' : _this$props$content,
          noHtml = _this$props.noHtml,
          noHtmlExceptMatchers = _this$props.noHtmlExceptMatchers,
          disableLineBreaks = _this$props.disableLineBreaks,
          disableWhitelist = _this$props.disableWhitelist,
          emptyContent = _this$props.emptyContent,
          parsedContent = _this$props.parsedContent;

      if (parsedContent) {
        return parsedContent;
      }

      var markup = new _Parser.default(content || '', {
        disableLineBreaks: disableLineBreaks,
        disableWhitelist: disableWhitelist,
        noHtml: noHtml,
        noHtmlExceptMatchers: noHtmlExceptMatchers
      }).parse();
      return markup.length > 0 ? markup : emptyContent;
    }
  }, {
    key: "render",
    value: function render() {
      var content = this.getContent();
      var tag = this.props.tagName;
      return tag === 'fragment' ? _react.default.createElement(_react.default.Fragment, null, content) : _react.default.createElement(_Element.default, {
        tagName: tag || 'div'
      }, content);
    }
  }]);
  return Markup;
}(_react.default.PureComponent);

exports.default = Markup;
(0, _defineProperty2.default)(Markup, "propTypes", {
  content: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.oneOf([null])]),
  disableLineBreaks: _propTypes.default.bool,
  disableWhitelist: _propTypes.default.bool,
  emptyContent: _propTypes.default.node,
  noHtml: _propTypes.default.bool,
  noHtmlExceptMatchers: _propTypes.default.bool,
  parsedContent: _propTypes.default.node,
  tagName: _propTypes.default.oneOfType([_propTypes.default.any, _propTypes.default.oneOf(['fragment'])])
});
(0, _defineProperty2.default)(Markup, "defaultProps", {
  content: '',
  disableLineBreaks: false,
  disableWhitelist: false,
  emptyContent: null,
  noHtml: false,
  noHtmlExceptMatchers: false,
  parsedContent: null,
  tagName: 'div'
});
