// interweave/lib/Matcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var Matcher = function () {
  function Matcher(name) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var factory = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    (0, _classCallCheck2.default)(this, Matcher);
    (0, _defineProperty2.default)(this, "options", void 0);
    (0, _defineProperty2.default)(this, "propName", void 0);
    (0, _defineProperty2.default)(this, "inverseName", void 0);
    (0, _defineProperty2.default)(this, "factory", void 0);

    if ("production" !== process.env.NODE_ENV) {
      if (!name || name.toLowerCase() === 'html') {
        throw new Error("The matcher name \"".concat(name, "\" is not allowed."));
      }
    }

    this.options = (0, _objectSpread2.default)({}, options);
    this.propName = name;
    this.inverseName = "no".concat(name.charAt(0).toUpperCase() + name.slice(1));
    this.factory = factory;
  }

  (0, _createClass2.default)(Matcher, [{
    key: "createElement",
    value: function createElement(match, props) {
      var element = null;

      if (this.factory) {
        element = _react.default.createElement(this.factory, props, match);
      } else {
        element = this.replaceWith(match, props);
      }

      if ("production" !== process.env.NODE_ENV) {
        if (typeof element !== 'string' && !_react.default.isValidElement(element)) {
          throw new Error("Invalid React element created from ".concat(this.constructor.name, "."));
        }
      }

      return element;
    }
  }, {
    key: "doMatch",
    value: function doMatch(string, pattern, callback) {
      var matches = string.match(pattern instanceof RegExp ? pattern : new RegExp(pattern, 'i'));

      if (!matches) {
        return null;
      }

      return (0, _objectSpread2.default)({}, callback(matches), {
        match: matches[0]
      });
    }
  }, {
    key: "onBeforeParse",
    value: function onBeforeParse(content, props) {
      return content;
    }
  }, {
    key: "onAfterParse",
    value: function onAfterParse(content, props) {
      return content;
    }
  }]);
  return Matcher;
}();

exports.default = Matcher;
